File: sigrun_lua/API/items/MenuCheckboxItem.lua
---@class MenuCheckboxItem : MenuItem
---@field public _Checked boolean Internal state of the checkbox
---@field public CheckBoxStyle number Style index for the checkbox (e.g., cross, tick)
---@field public OnCheckboxChanged fun(tab: BaseTab, item: MenuCheckboxItem, checked: boolean) Callback triggered when the state changes
MenuCheckboxItem = {}
MenuCheckboxItem.__index = MenuCheckboxItem
setmetatable(MenuCheckboxItem, { __index = MenuItem })
MenuCheckboxItem.__call = function() return "MenuCheckboxItem" end

---Creates a new MenuCheckboxItem instance.
---@param Text string The item label
---@param Check boolean Initial checked state
---@param checkStyle number|nil Optional style index
---@param Description string|table Description text or table
---@param color SColor|nil Custom background color
---@param highlightColor SColor|nil Custom highlight color
---@return MenuCheckboxItem
function MenuCheckboxItem.New(Text, Check, checkStyle, Description, color, highlightColor)
    local base = MenuItem.New(Text or "", Description or "", color, highlightColor)
    base._Checked = ToBool(Check)
    base.CheckBoxStyle = checkStyle or 0
    base.ItemId = 2 -- Fixed ID for Checkbox items in Scaleform
    
    -- Default callback placeholder
    base.OnCheckboxChanged = function(tab, item, checked) end
    
    return setmetatable(base, MenuCheckboxItem)
end

---Gets or sets the checked state of the item.
---@param bool boolean|nil New state to set
---@return boolean|nil Current state if no parameter is provided
function MenuCheckboxItem:Checked(bool)
    if bool ~= nil then
        self._Checked = ToBool(bool)
        if self.ParentColumn ~= nil then
            local it = IndexOf(self.ParentColumn.Items, self)
            self.ParentColumn:SendItemToScaleform(it, true)
        end
    else
        return self._Checked
    end
end

---------------------------
-- Unsupported Methods
---------------------------
-- These methods are overridden to prevent usage on Checkbox items 
-- as they are not supported by the Scaleform layout.

function MenuCheckboxItem:RightLabelFont(itemFont)
    error("MenuCheckboxItem does not support a right label")
end

function MenuCheckboxItem:RightBadge()
    error("MenuCheckboxItem does not support right badges")
end

function MenuCheckboxItem:RightLabel()
    error("MenuCheckboxItem does not support a right label")
end

function MenuCheckboxItem:SetSubColumn(itemListColumn, newTitle, showArrow, hideTabs)
    error("MenuCheckboxItem does not support subcolumns by nature")
end

File: sigrun_csharp/SigrunClient/API/Items/MenuCheckboxItem.cs
using SigrunClient.API.Elements;
using System;

namespace SigrunClient.API.Items
{
    public enum MenuCheckboxStyle
    {
        Cross,
        Tick
    }

	public delegate void ItemCheckboxEvent(MenuCheckboxItem sender, bool Checked);
	public class MenuCheckboxItem : MenuItem
    {
        private bool _checked;

        /// <summary>
        /// Triggered when the checkbox state is changed.
        /// </summary>
        public event ItemCheckboxEvent CheckboxEvent;

        public MenuCheckboxStyle Style { get; }

        /// <summary>
        /// Checkbox item with a toggleable checkbox.
        /// </summary>
        /// <param name="text">Item label.</param>
        /// <param name="check">Boolean value whether the checkbox is checked.</param>
        public MenuCheckboxItem(string text, bool check) : this(text, check, "")
        {
        }

        /// <summary>
        /// Checkbox item with a toggleable checkbox.
        /// </summary>
        /// <param name="text">Item label.</param>
        /// <param name="check">Boolean value whether the checkbox is checked.</param>
        /// <param name="description">Description for this item.</param>
        public MenuCheckboxItem(string text, bool check, string description) : this(text, MenuCheckboxStyle.Tick, check, description, SColor.HUD_Pause_bg, SColor.HUD_White)
        {
        }

        /// <summary>
        /// Checkbox item with a toggleable checkbox.
        /// </summary>
        /// <param name="text">Item label.</param>
        /// <param name="style">CheckBox style (Tick or Cross).</param>
        /// <param name="check">Boolean value whether the checkbox is checked.</param>
        /// <param name="description">Description for this item.</param>
        public MenuCheckboxItem(string text, MenuCheckboxStyle style, bool check, string description) : this(text, style, check, description, SColor.HUD_Pause_bg, SColor.HUD_White)
        {
        }

        /// <summary>
        /// Checkbox item with a toggleable checkbox.
        /// </summary>
        /// <param name="text">Item label.</param>
        /// <param name="style">CheckBox style (Tick or Cross).</param>
        /// <param name="check">Boolean value whether the checkbox is checked.</param>
        /// <param name="description">Description for this item.</param>
        /// <param name="mainColor">Main item color.</param>
        /// <param name="highlightColor">Highlight item color.</param>
        public MenuCheckboxItem(string text, MenuCheckboxStyle style, bool check, string description, SColor mainColor, SColor highlightColor) : base(text, description, mainColor, highlightColor)
        {
            Style = style;
            _checked = check;
            _itemId = 2;
        }


        /// <summary>
        /// Change or get whether the checkbox is checked.
        /// </summary>
        public bool Checked
        {
            get => _checked;
            set
            {
                _checked = value;
                if (ParentColumn != null && ParentColumn.visible)
                    ParentColumn.SendItemToScaleform(ParentColumn.Items.IndexOf(this), true);
            }
        }

        public void CheckboxEventTrigger()
        {
            CheckboxEvent?.Invoke(this, Checked);
        }

        public override void SetRightBadge(string txd, string txn, SColor mainColor, SColor highlightColor)
		{
            throw new Exception("MenuCheckboxItem cannot have a right badge.");
        }

        public override void SetRightLabel(string text)
        {
            throw new Exception("MenuCheckboxItem cannot have a right label.");
        }
    }
}